src/TypeFormatter/ReferenceTypeFormatter.ts

```diff
@@ -16,8 +17,12 @@ export class ReferenceTypeFormatter implements SubTypeFormatter {
         return { $ref: `#/definitions/${this.encodeRefs ? encodeURIComponent(ref) : ref}` };
     }
     public getChildren(type: ReferenceType): BaseType[] {
-        if (type.getType() instanceof DefinitionType) {
-            return [];
+        const referredType = type.getType();
```

Comment:
Does `derefType(...)` work here?

Reply:
I think it goes too deep in the hierarchy. It returns the `UnionType` but we need the `DefinitionType` that wraps it.

Reply:
Are you saying we need the last definition type right above the non definition type? Maybe it's with pulling that functionality into a until function. We could add a flag to the deref function.

Reply:
So, we have a piece of path from a RefType to a UnionType:

`ReferenceType` -> `DefinitionType` -> `AliasType` -> `UnionType`

Originally the traversal didn't reach the DefinitionType or its children, but we need to return a/the DefinitionType that wraps the UnionType. The intervening AliasType is a bit mystery to me. I guess we could also return an instance of a new DefinitionType that wraps the the UnionType directly(?) Anyway, I'm not sure which are exactly the situations where the traversal in ReferenceTypeFormatter should proceed to the DefinitionType. Now I used the presence of an AliasType as an indication for that. Although it works, it may not make much sense.

It actually seems that we could always traverse the DefinitionType. Tests seem to pass. I just thought that there was some reason for stopping propagation there. Maybe it was again an optimization that isn't needed any more?

Reply:
If we can just not have this check here, let's delete it. I don't understand the logic otherwise ;-)

Reply:
If I delete the block starting with: `if (referredType instanceof DefinitionType) {` or the original `if (type.getType() instanceof DefinitionType) {`, two tests fail:

The one written by me (`type-intersection-union-recursive-interface`):

```
      "Union": Object {
-       "anyOf": Array [
-         Object {
-           "$ref": "#/definitions/Container",
-         },
-         Object {
-           "$ref": "#/definitions/Silly",
-         },
-       ],
+       "$ref": "#/definitions/Union",
      },
```

... and a pretty much identical case in `vega-lite.test.ts`:

```
      "RepeatSpec": Object {
-       "anyOf": Array [
-         Object {
-           "$ref": "#/definitions/NonLayerRepeatSpec",
-         },
-         Object {
-           "$ref": "#/definitions/LayerRepeatSpec",
-         },
-       ],
+       "$ref": "#/definitions/RepeatSpec",
      },
```

A silly short-circuited circular reference is produced.

To be honest, I'm a bit puzzled why returning an empty array prevents that. However, the original code that just returned `[]` for DefinitionTypes fails with the new test case (and #1046).

Reply:
Hmm, that comes from a long time ago: https://github.com/vega/ts-json-schema-generator/commit/dd6b1bc88c0ceb5292d109d58186c67d1c7f827e. I don't know.

Reply:
Hmm I spent some time stepping through the execution but didn't see the obvious bug yet. I added a breakpoint in `ReferenceTypeFormatter.ts` and then debugged your new test case and the existing test cases. I then set the code to generate the schema files from the tests and disabled the reachability check.

Alternatively, I created a test file and then ran `yarn run run -p test.ts -t Intersection`

I'm not satisfied with the hack above tbh. 

The alias type makes sense. It's the union. 

<img width="1273" alt="Screen Shot 2022-01-13 at 18 59 50" src="https://user-images.githubusercontent.com/589034/149427715-50f15ea0-1757-434b-ad91-0f3806c9c662.png">

It goes away if you inline the type. 

I would really like to understand why we don't format the definitions for the children.

Reply:
I have a hunch what's happening.

We can expand the Intersection type to `export type Intersection = (A & C) | (B & C);`. See https://github.com/vega/ts-json-schema-generator/pull/985 and https://github.com/vega/ts-json-schema-generator/commit/d9b9a15aa5849a2e298c009f60ada0d3539b6f47. 

This means, we think that we don't need the Union type anymore, since we transformed it in https://github.com/vega/ts-json-schema-generator/commit/d9b9a15aa5849a2e298c009f60ada0d3539b6f47#diff-79b658f531e0da963f97358d1c4ab65d18674a91e460f86500cca7cbdee966b3R23. 

We addressed part of this for aliases that are inside the union but not for the union itself in https://github.com/vega/ts-json-schema-generator/pull/985/files.

So I think the fix needs to be in the Intersection/Union code, not the reference type.

Reply:
We return [] because we know that definitions should already have been created. They usually already are but here is the exception.

Reply:
Hmm, or maybe we can just return `this.childTypeFormatter.getChildren(referredType)` instead of doing the check. Maybe we do this because we think the definition is already through some other way defined. However, since we do caching and detect duplicated, I think we should be fine just returning the children.

Reply:
Thank you for spending time on this issue/pr!